src/vsr/client_replies.zig

```diff
@@ -0,0 +1,342 @@
+// the same client will queue another request. If they do (within the same checkpoint),
+// then we no longer need to persist the original reply.
+pub fn ClientRepliesType(comptime Storage: type) type {
+    return struct {
+        const ClientReplies = @This();
+
+        const Read = struct {
+            client_replies: *ClientReplies,
+            completion: Storage.Read,
+            callback: fn (client_replies: *ClientReplies, reply: ?*Message) void,
+            message: *Message,
+            /// The header of the expected reply.
+            header: vsr.Header,
+        };
+
+        const Write = struct {
+            client_replies: *ClientReplies,
+            completion: Storage.Write,
+            slot: Slot,
+            message: *Message,
+        };
+
+        storage: *Storage,
+        message_pool: *MessagePool,
+        replica: u8,
+
+        reads: IOPS(Read, constants.client_replies_iops_read_max) = .{},
+        writes: IOPS(Write, constants.client_replies_iops_write_max) = .{},
+
+        /// Track which slots have a write currently in progress.
+        writing: std.StaticBitSet(constants.clients_max) =
+            std.StaticBitSet(constants.clients_max).initEmpty(),
+
+        /// Guard against multiple concurrent writes to the same slot.
+        /// Pointers are into `writes`.
+        write_queue: RingBuffer(*Write, constants.client_replies_iops_write_max, .array) = .{},
+
+        ready_next_tick: Storage.NextTick = undefined,
+        ready_callback: ?fn (*ClientReplies) void = null,
+
+        checkpoint_next_tick: Storage.NextTick = undefined,
+        checkpoint_callback: ?fn (*ClientReplies) void = null,
+
+        pub fn init(storage: *Storage, message_pool: *MessagePool, replica_index: u8) ClientReplies {
+            return .{
+                .storage = storage,
+                .message_pool = message_pool,
+                .replica = replica_index,
+            };
+        }
+
+        pub fn deinit(client_replies: *ClientReplies) void {
+            {
+                var it = client_replies.reads.iterate();
+                while (it.next()) |read| client_replies.message_pool.unref(read.message);
+            }
+            {
+                var it = client_replies.writes.iterate();
+                while (it.next()) |write| client_replies.message_pool.unref(write.message);
+            }
+            // Don't unref `write_queue`'s Writes — they are a subset of `writes`.
+        }
+
+        pub fn read_reply(
+            client_replies: *ClientReplies,
+            slot: Slot,
+            entry: *const Entry,
+            callback: fn (*ClientReplies, ?*Message) void,
+        ) void {
+            const client = entry.header.client;
+
+            if (client_replies.writing.isSet(slot.index)) {
+                var writes = client_replies.writes.iterate();
+                var write_latest: ?*const Write = null;
+                while (writes.next()) |write| {
```

Comment:
Hm, shouldn't we have a symmetric in loop `write_reply` (or at least an assert)?

What happens if we schedule fall through here, schedule a read, and then try to wirte to the same slot before the read is completed?